cleanup + provide auth logic

laur89 opened the pull request:
- fixes #2 (almost - no album selection);
- code cleanup;
    - syntax closer to bash (as opposed to POSIX);
    - localise variables;
    - all global vars/constants in upper case;
    - tabs->spaces;
- add possibility to upload/delete pictures with authenticated user;
- add possibility to automatically delete uploaded pictures;
- add possibility to upload to albums;
- add external configuration;
- add notifications;
- add logging;
- remove regex json parsing - introduce dependency on jq;

I understand this is a substantial change, but everything _should_ be backwards compatible.
Can anyone test in mac please?

alecmev commented:
Basically a rewrite :smile: Thanks for the effort!

Don't have a Mac to test on, sorry. Just Linux.

alecmev commented on imgurbash2:
```diff
@@ -22,216 +22,599 @@
+    TOKEN_EXPIRE_TIME=0  # set default, in case it can't be found in $CREDENTIALS_FILE yet
+
+    [[ -s "$CREDENTIALS_FILE" ]] && source "$CREDENTIALS_FILE"
+    check_oauth2_client_secrets
```
Initially, there's no `REFRESH_TOKEN` (Imgur doesn't give you one when you register an app), so this fails on the first use.

laur89 replied:
But it printed the instructions in console, correct?

Would prefer for users to perform the initial setup manually for simplicity's sake.

alecmev replied:
Wait, what's the non-manual way?

Anyway, yes, it does print the message, but Imgur app registration UI gives you only `CLIENT_ID` and `CLIENT_SECRET`, there's no `REFRESH_TOKEN`, and this check fails when it doesn't find it in the config.

laur89 replied:
Guess I need to re-visit the app registration. Last time I followed that link printed in console, I did end up with refresh token as well. Might've taken additional steps and later forgot about it.

alecmev replied:
[This form](https://api.imgur.com/oauth2/addclient), right?

alecmev replied:
Just in case, `imgur-screenshot` works without a pre-existing refresh token.

laur89 replied:
It's not that easy.
Basically yes, that form gives us `client_id` & `client_secret`, which then need to be used over auth api to receive `refresh_token` (via https://api.imgur.com/oauth2/authorize?client_id=YOUR_CLIENT_ID&response_type=token)

Note imgur-screenshot uses PIN auth (response_type=pin), which is deprecated.

Added a bit more descriptive err message, but bottom line is users need to get it sorted manually beforehand. If it can be improved, I'm all ears.

alecmev replied:
Dang, didn't know about the PIN auth deprecation.

Maybe at least make the process interactive? First, ask users to paste in their app `client_id`, and then generate a link they need to follow, and then ask them to copy the redirect URL and paste it back into the app, so then `imgurbash2` can parse it programmatically. More work though.

laur89 replied:
Not a bad idea. i'll give it a go.

laur89 commented on imgurbash2:
```diff
@@ -22,216 +22,599 @@
+    if [[ -f "$file" ]]; then
+        hdrs+=('-F' "image=@$file")
+    elif is_valid_url "$file"; then
+        # TODO: only works when album_id not defined?
```
Could anyone confirm logic for external image upload is valid?
For some reason can't get it to work if we want to upload to a specific album.

ram-on commented:
Hello @laur89.  Thanks !  I'll test it tomorrow on a macOs.  In the meantime case you change from spaces to tabs - I just like tabs better.

laur89 commented:
@ram-on can you check on mac? If the auth changes look ok to you, i'd rather squash it all to single commit.

ram-on commented:
@laur89 Apologies for the late reply.  Thank you for the PR!  The following are issues that I've noted:

1. Change the first line of the script from `#!/bin/bash` to `#!/usr/bin/env bash`  This is because macOS' default bash version is ver 3 and hence we need to use the bash version installed via `brew` -- i.e. both bash 3 and bash 4 can be simultaneously available on macOS (refer to https://github.com/ram-on/imgurbash2/commit/ddf43d94ffec091135b4ede8f24d81b327c1d469)

1. This script can no longer run on bash version 3.  Update the code so that it will not run if bash version 3 is available and the user is notified accordingly.

1. I would like to have this script independent to any other libraries.  This include the `jq` library which can be easily substituted by two one-liner functions (refer to https://github.com/ram-on/imgurbash2/commit/ddf43d94ffec091135b4ede8f24d81b327c1d469).  Please remove any reference to `jq`.

1. With respect to `notify()` function:  Please ensure that `growlnotify`, `terminal-notifier` and `notify-send` commands are all optional.

1. Regarding the auto-delete:  pls update the documentation (readme) such that it is clear that the auto-delete is performed by this script while running in the background.  Also, update the script's help as on BSD's toolkit, the sleep command does not accept any suffix - the supplied interval is in seconds...

1. Given that there are out there scripts that depend on imgurbash2, please make sure that the output format is reverted back to what was before.  I.e. Output format should be as follows:  `https://i.imgur.com/XXXXXXXXX.jpg    (Delete Hash = ZZZZZZZZZ)`

1. Bug: I'm getting errors when executing this command: `./imgurbash2 -l true ~/testimg.png`:  `.config/imgurbash2/credentials.conf: No such file or directory` on lines 483, 489 and 568.  Solution would be to check if the .config and imgurbash2 dirs are created (if not create them).

1. Bug:  After manually creating ".config/imgurbash2" directory, I've executed `./imgurbash2 -l true ~/testimg.png` and filled in the required fields.  After that, I've got this error:
```
request failed with status [400]: {
  "code": 403,
  "message": "Forbidden",
  "type": "ImgurException",
  "exception": []
}
```

1. Bug:  if you upload multiple images, the multiple imgur's URLs are not all copied to clipboard, e.g.: for the command `imgurbash2 hello.png qwerty.png`, the last clipboard **entry** should be:
```
https://i.imgur.com/AAAAAAA.jpg
https://i.imgur.com/BBBBBBB.png
```
Instead, just `https://i.imgur.com/BBBBBBB.png` is being copied to the clipboard

laur89 commented:
Excellent review. Some notes:

 `3.` don't quite agree with parsing json with flaky regex instead of dedicated and well-known tools, but will fold on this one.

 `6.` what about the message about automatic deletion? Wouldn't this additional message potentially corrupt clients' parsers? Note I reverted the default success-case atm, but not sure if that's enough.

 `8.` no idea what's the case here; did you confirm credentials had been correctly been parsed into `credentials.conf`? Replaced perl expression with sed for the time being.

 `9.` is feature not a bug; don't see much value in having all the urls as single clipboard entry - easier to use if they're separate entries on your clipboard. This of course assumes usage of a clipboard manager.
Sure you want all the urls as a single entry?
Suspect you won't agree to this, so [post-review 2](https://github.com/laur89/imgurbash2/commit/f5be0c9b8e179ca7dab5c3c14a08dc39e874e50f) contains change to copy all urls as a single entry.

ram-on commented:
### Error 1:

Hi @laur89, I'm getting an error when executing this command:
```
./imgurbash2 -l true abc.png
```

Error occures when I'm asked to input the "redirect url":
```
redirect url: https://api.imgur.com/oauth2/authorize?client_id=XXXXXX&response_type=token
./imgurbash2: line 614: https://api.imgur.com/oauth2/authorize?client_id=XXXXXX&response_type=token: syntax error in expression (error token is "://api.imgur.com/oauth2/authorize?client_id=XXXXXX&response_type=token")
imgurbash2: request failed with status [403]: The access token provided is invalid.
imgurbash2: uploading [abc.png] failed
```

-----------------

### Error 2:
When I've used my old credentials file and when executing this command ...:
```
./imgurbash2 -l true ~/testimg.png
```

... this error message was displayed:
```
imgurbash2: request failed with status [400]: {"data":{"error":{"code":403,"message":"Forbidden","type":"ImgurException","exception":[]},"request":"\/3\/image","method":"POST"},"success":false,"status":400}
imgurbash2: uploading [/Users/ramon/testimg.png] failed
```

Can you pls look into these errors?

laur89 commented:
Err 1 seems to be resulting from an incorrect redirect url being pasted in.
The redirect url should be something like `domainYouChose/#access_token=1234abc&expires_in=315360000&token_type=bearer&refresh_token=dedb3822&account_username=ram-on&account_id=83838383`

Looks like the auth url (`https://api.imgur.com/oauth2/authorize?client_id=jejej38383&response_type=token`) was pasted instead.

Err 2 is trickier to debug. For both errors, it would be nice to see the contents of the parsed `credentials.conf` (with sensitive data redacted). Expected contents should look similar to

```
CLIENT_ID='g9072135f47bc38'
CLIENT_SECRET='e10468480b444ff37f36085891f918770ab0ed2b'
ACCESS_TOKEN='r48c938b66660acfc5z43391d54ce1fb95e8a94'
REFRESH_TOKEN='18f3c062e6ad4397876ahj5c0c8e7833b1vc0a674'
TOKEN_EXPIRE_TIME='1845762520'

laur89 commented:
@ram-on can you please take a look again? Don't have access to apple products so sadly can't be much more of a help at debugging issues on macos.

ram-on commented:
@laur89 Sorry for the delay!  I've tested it out, bug no. 1 is still there.  I've descided to debug it and found out that the `_parse()` of line 536 was not extracting the "access_token", "refresh_token" and "expires_in".  The correct code for `_parse()` should be as follows:

```bash
_parse() {
	sed -e 's#.*'$1'=\([^&]*\).*#\1#' <<< "$url"
}
```

Please ensure that the above `_parse()` works well on the GNU userland and then update your branch accordingly.

However, after the above tokens are correctly expracted by using the above `_parse()` function, I'm still getting the below error:

```
imgurbash2: request failed with status [400]: {"data":{"error":{"code":403,"message":"Forbidden","type":"ImgurException","exception":[]},"request":"\/3\/image","method":"POST"},"success":false,"status":400}
imgurbash2: uploading [/Users/ramon/testimg2.jpg] failed
```

I'm also getting this error on FreeBSD.  Can you take at look?

laur89 commented:
What are the contents of parsed `credentials.conf` when you get the 403 from imgur?

ram-on commented:
```
CLIENT_ID='15 AlphaNumeric characters'
CLIENT_SECRET='40 AlphaNumeric characters'
ACCESS_TOKEN='40 AlphaNumeric characters'
REFRESH_TOKEN='40 AlphaNumeric characters'
TOKEN_EXPIRE_TIME='10 digits'
```

laur89 commented:
So missing cred data is not causing this issue.
Can you run this directly from the shell to rule out curl option incompatibilities:

	declare -a hdrs=('-X' POST)
	hdrs+=('-H' "Authorization: Bearer ${ACCESS_TOKEN}")
    hdrs+=('-F' "image=@$FILE")
	[[ -n "$TITLE" ]] && hdrs+=('-F' "title=$TITLE")
	[[ -n "$ALBUM_ID" ]] && hdrs+=('-F' "album=$ALBUM_ID")

    API_CALL_CONNECT_TIMEOUT_SEC=5
    API_CALL_TIMEOUT_SEC=60
    API_CALL_RETRIES=1
	curl -v --compressed \
			--connect-timeout "$API_CALL_CONNECT_TIMEOUT_SEC" \
			-m "$API_CALL_TIMEOUT_SEC" \
			--retry "$API_CALL_RETRIES" \
			--stderr - \
			-sSL \
			"${hdrs[@]}" \
			"https://api.imgur.com/3/image"

Don't forget to set `ACCESS_TOKEN`, `FILE`, and optionally `ALBUM_ID` & `TITLE`

ram-on commented:
@laur89 Hi, please find the log for the above code here:  [log.txt](https://github.com/ram-on/imgurbash2/files/2195663/log.txt).  FYI, I've just replaced the access token string with "XXXXXXXX...".

My curl version:
```
curl 7.54.0 (x86_64-apple-darwin17.0) libcurl/7.54.0 LibreSSL/2.0.20 zlib/1.2.11 nghttp2/1.24.0
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz HTTP2 UnixSockets HTTPS-proxy
```

laur89 commented:
Running out of ideas here... assuming access_token is correct, I see no reason why it wouldn't work.
Just for the sake of it, could you try
- removing `--connect-timeout`, `--retry` and `-m` options from curl cmd;
- upgrading curl to latest;

Can you confirm anon upload with the script works?

alecmev commented:
As a data point, I've been using this for two months now, zero issues, zero manual interventions.

laur89 commented:
@jeremejevs that's because all the remaining issues appear to be macos-specific.

ram-on commented:
@laur89 I've just setup a linux box and I'm still getting this error:
```
imgurbash2: request failed with status [400]: {"data":{"error":{"code":403,"message":"Forbidden","type":"ImgurException","exception":[]},"request":"\/3\/image","method":"POST"},"success":false,"status":400}
imgurbash2: uploading [/home/ramon/aa.jpg] failed
```

Hence, the error must be in how the key is being generated via imgur?

Please detail every step how did you generate the key (e.g. what Authorization type did you select?).  To be honest I've tried both "OAuth 2 authorization with a callback URL" and "OAuth 2 authorization without a callback URL". Neither option worked on both Linux and macOS...

ram-on commented:
@laur89 Gentle reminder pls.

alecmev commented:
@ram-on I reset my credentials and went through the whole procedure, worked fine. Steps 1-6 do need to be documented somewhere though, at least in the readme.

1. Go to https://api.imgur.com/oauth2/addclient
2. Select "OAuth 2 authorization with a callback URL".
3. Set callback URL to https://imgur.com (that's what the URL of my several years old apps is)
4. Submit.
5. Open `~/.config/imgurbash2/credentials.conf`.
6. Set `CLIENT_ID` and `CLIENT_SECRET` to the values displayed on that app registration page.
7. Proceed as directed by `imgurbash2 --login true <image>`.

@laur89 While writing this down, found a bug: if `credentials.conf` doesn't have a trailing line break while having only `CLIENT_ID` and `CLIENT_SECRET`, then the `ACCESS_TOKEN` line will get concatenated with the last line in the file, ruining further uploads.

ram-on commented:
@jeremejevs Follow your steps -- yet no luck at all.  I'm running this on a linux machine.  I'm still gettig the below error:
```
request failed with status [400]: {
  "code": 403,                           
  "message": "Forbidden",        
  "type": "ImgurException",       
  "exception": []                         
} 
```

Any tips?

alecmev commented:
@ram-on Can you try changing your Imgur email address? A few ShareX users have experienced something very similar: https://github.com/ShareX/ShareX/issues/2962

ram-on commented:
@jeremejevs Wow that did the trick!

alecmev commented:
What a ridiculous issue :smile: Maybe it's time to start self-hosting...

More on topic, glad to see this merged! Thanks for the work, @laur89.

ram-on commented:
Indeed!  Thanks @laur89

laur89 commented:
Sorry, been on hols and otherwise busy.

Happy to see y'all got it sorted, cheers!
